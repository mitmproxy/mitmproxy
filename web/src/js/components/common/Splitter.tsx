import React, { Component } from "react";
import ReactDOM from "react-dom";
import classnames from "classnames";

type SplitterState = {
    applied: boolean;
    startX: number;
    startY: number;
};

type SplitterProps = {
    axis: string;
};

export default class Splitter extends Component<SplitterProps, SplitterState> {
    static defaultProps = { axis: "x" };

    constructor(props, context) {
        super(props, context);

        this.state = { applied: false, startX: 0, startY: 0 };

        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseDown = this.onMouseDown.bind(this);
        this.onMouseUp = this.onMouseUp.bind(this);
        this.onDragEnd = this.onDragEnd.bind(this);
    }

    onMouseDown(e) {
        this.setState({ startX: e.pageX, startY: e.pageY });

        window.addEventListener("mousemove", this.onMouseMove);
        window.addEventListener("mouseup", this.onMouseUp);
        // Occasionally, only a dragEnd event is triggered, but no mouseUp.
        window.addEventListener("dragend", this.onDragEnd);
    }

    onDragEnd() {
        ReactDOM.findDOMNode(this).style.transform = "";

        window.removeEventListener("dragend", this.onDragEnd);
        window.removeEventListener("mouseup", this.onMouseUp);
        window.removeEventListener("mousemove", this.onMouseMove);
    }

    onMouseUp(e) {
        this.onDragEnd();

        const node = ReactDOM.findDOMNode(this);
        const prev = node.previousElementSibling;

        let flexBasis = prev.offsetHeight + e.pageY - this.state.startY;

        if (this.props.axis === "x") {
            flexBasis = prev.offsetWidth + e.pageX - this.state.startX;
        }

        prev.style.flex = `0 0 ${Math.max(0, flexBasis)}px`;
        node.nextElementSibling.style.flex = "1 1 auto";

        this.setState({ applied: true });
        this.onResize();
    }

    onMouseMove(e) {
        let dX = 0;
        let dY = 0;
        if (this.props.axis === "x") {
            dX = e.pageX - this.state.startX;
        } else {
            dY = e.pageY - this.state.startY;
        }
        ReactDOM.findDOMNode(
            this
        ).style.transform = `translate(${dX}px, ${dY}px)`;
    }

    onResize() {
        // Trigger a global resize event. This notifies components that employ virtual scrolling
        // that their viewport may have changed.
        window.setTimeout(
            () => window.dispatchEvent(new CustomEvent("resize")),
            1
        );
    }

    reset(willUnmount) {
        if (!this.state.applied) {
            return;
        }

        const node = ReactDOM.findDOMNode(this);

        if (node.previousElementSibling) {
            node.previousElementSibling.style.flex = "";
        }
        if (node.nextElementSibling) {
            node.nextElementSibling.style.flex = "";
        }

        if (!willUnmount) {
            this.setState({ applied: false });
        }
        this.onResize();
    }

    componentWillUnmount() {
        this.reset(true);
    }

    render() {
        return (
            <div
                className={classnames(
                    "splitter",
                    this.props.axis === "x" ? "splitter-x" : "splitter-y"
                )}
            >
                <div onMouseDown={this.onMouseDown} draggable="true" />
            </div>
        );
    }
}
