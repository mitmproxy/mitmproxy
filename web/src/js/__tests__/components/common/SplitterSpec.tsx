import * as React from "react";
import Splitter from "../../../components/common/Splitter";
import { act, render } from "../../test-utils";

describe.each([["x"], ["y"]])("Splitter Component", (axis) => {
    it(`should render correctly (${axis} axis)`, () => {
        const ref = React.createRef<Splitter>();
        const { asFragment, unmount } = render(
            <>
                <div></div>
                <Splitter axis={axis} ref={ref} />
                <div></div>
            </>,
        );
        const splitter = ref.current!;

        expect(asFragment()).toMatchSnapshot();

        act(() => {
            splitter.onPointerDown({
                target: {
                    setPointerCapture: () => 0,
                } as unknown,
                pageX: 100,
                pageY: 200,
                pointerId: 42,
            } as React.PointerEvent<HTMLDivElement>);
        });
        expect(splitter.state.startPos).toBe(axis === "x" ? 100 : 200);

        act(() => {
            splitter.onPointerMove({
                pageX: 300,
                pageY: 300,
                pointerId: 42,
            } as React.PointerEvent<HTMLDivElement>);
        });
        expect(splitter.node.current!.style.transform).toBeTruthy();

        act(() => {
            splitter.onLostPointerCapture({
                pageX: 400,
                pageY: 400,
                pointerId: 42,
            } as React.PointerEvent<HTMLDivElement>);
        });
        expect(asFragment()).toMatchSnapshot();

        unmount();
    });
});
