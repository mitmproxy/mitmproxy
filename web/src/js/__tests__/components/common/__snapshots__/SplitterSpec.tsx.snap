// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter Component should render correctly (x axis) 1`] = `
<DocumentFragment>
  <div />
  <div
    class="splitter splitter-x"
  >
    <div />
  </div>
  <div />
</DocumentFragment>
`;

exports[`Splitter Component should render correctly (x axis) 2`] = `
<DocumentFragment>
  <div
    style="flex: 0 0 300px;"
  />
  <div
    class="splitter splitter-x"
    style=""
  >
    <div />
  </div>
  <div
    style="flex: 1 1 auto;"
  />
</DocumentFragment>
`;

exports[`Splitter Component should render correctly (y axis) 1`] = `
<DocumentFragment>
  <div />
  <div
    class="splitter splitter-y"
  >
    <div />
  </div>
  <div />
</DocumentFragment>
`;

exports[`Splitter Component should render correctly (y axis) 2`] = `
<DocumentFragment>
  <div
    style="flex: 0 0 200px;"
  />
  <div
    class="splitter splitter-y"
    style=""
  >
    <div />
  </div>
  <div
    style="flex: 1 1 auto;"
  />
</DocumentFragment>
`;
